akka-http-core/src/main/scala/akka/http/scaladsl/model/ErrorInfo.scala

```diff
@@ -69,7 +71,7 @@ object ErrorInfo {
 }
 
 /** Marker for exceptions that provide an ErrorInfo */
-abstract class ExceptionWithErrorInfo(val info: ErrorInfo, cause: Throwable) extends RuntimeException(info.formatPretty, cause) {
+abstract class ExceptionWithErrorInfo(val info: ErrorInfo, cause: Throwable) extends RuntimeException(info.format(withDetail = false), cause) {
```

Comment:
This is tricky indeed.

Ideally, we should not output 'possibly sensitive' information to clients unless verbose `verbose-error-messages` is on. Without further information we should consider the `message` of an exception as 'possibly sensitive' and not output it, so from that perspective `withDetail = true` would be reasonable.

In the case of EntityStreamSizeException, this exception is not caught by the exception handler, but turned into a `MalformedRequestContentRejection`. Since rejections are meant to be shown to the client (and not necessarily go into the logging), I guess the rejection message should not contain 'possibly sensitive' information.

Unfortunately FormFieldDirectives does just put the exception message into the rejection message. Perhaps we could follow the approach in MarshallingDirectives and look at `verbose-error-messages` when turning an exception that has an `ErrorInfo` into a rejection, like https://github.com/jrudolph/akka-http/pull/2 ?

Reply:
These are good points. I guess we should try to untangle a few related but distinct aspects of it to make some progress:

 * The particular issue in question about the EntitySize exception
 * Which information is allowed to be included to users?
 * Which information stack traces are allowed to carry and how they should be logged

> The particular issue in question about the EntitySize exception

This can and maybe should be solved in isolation.

> Which information is allowed to be included in message to users in general?

I think we have quite good guidelines on this. In doubt, be conservative and make the messages as generic as possible to avoid attack vectors.

Neither this PR nor your suggestion actual addresses the real problem here: 

We should never include a clause like this which trusts generic exception implementations to do the right thing:

```scala
        case Failure(x)                                  => reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, x))
```

If we want to fix something generically it should be these places where we shouldn't include any concrete information about an exception with `verbose-error-message = off`.

> Which information stack traces are allowed to carry and how they should be logged

There are different levels of sensitive data. Critically, secrets should never be included in any messages that might be logged. Other details should be allowed to be logged to help with debugging though not be exposed to users in responses.

Reply:
> Neither this PR nor your suggestion actual addresses the real problem here

I agree: they happen to solve the EntitySize situation specifically, but...

> We should never include a clause like this which trusts generic exception implementations to do the right thing:
> 
> ```scala
>         case Failure(x)                                  => reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, x))
> ```

Indeed this is the 'real problem'. Unfortunately this won't be such an easy change, as we'd have to review how *else* to convey the details of this error then...

> If we want to fix something generically it should be these places where we shouldn't include any concrete information about an exception with `verbose-error-message = off`.

:+1: 

> > Which information stack traces are allowed to carry and how they should be logged
> 
> There are different levels of sensitive data. Critically, secrets should never be included in any messages that might be logged. Other details should be allowed to be logged to help with debugging though not be exposed to users in responses.

:+1:

Reply:
> Indeed this is the 'real problem'. Unfortunately this won't be such an easy change, as we'd have to review how _else_ to convey the details of this error then...

Yep, probably still use your suggestion for that and deal with the more generic things in another PR.

Reply:
> We should never include a clause like this which trusts generic exception implementations to do the right thing:
> 
> ```scala
>         case Failure(x)                                  => reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, x))
> ```
> 
> If we want to fix something generically it should be these places where we shouldn't include any concrete information about an exception with `verbose-error-message = off`.

On potential problem with this is that we use `IllegalArgumentExceptions` to transport parsing errors in various places. In those cases the details would now be hidden behind `verbose-error-messages` as well.

Reply:
Another aspect that I found, is that the processing varies between the `FormFieldDirectives`.

 * The `formField` variants use `toStrictEntity` which will fail with an exception directly without giving control to the form field processing at all, so the response is created from the exception handler ultimately.
 * The `formFieldSeq` / `Map` / `MultiMap` directives however do no (yet?) use `toStrictEntity` (because we didn't expect people to nest those with `formField` which might be to optimistic anyway). So those fail while actually running the unmarshaller so that the extra handling logic above is run to wrap the exceptions with rejections as we see fit.